uvicorn/middleware/proxy_headers.py

```diff
@@ -8,11 +8,18 @@
 
 https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers#Proxies
 """
-from typing import List
+from typing import List, Optional
+
+from asgiref.typing import (
+    ASGI3Application,
+    ASGIReceiveCallable,
+    ASGISendCallable,
+    Scope,
+)
 
 
 class ProxyHeadersMiddleware:
-    def __init__(self, app, trusted_hosts="127.0.0.1"):
+    def __init__(self, app: ASGI3Application, trusted_hosts: str = "127.0.0.1") -> None:
```

Comment:
I'm surprised by this, `trusted_hosts` is a list of strings or a string I think
I dont get why mypy allows that given the tests on that  module is using either a list or a single string.
I now realize we added the whole typing without the corresponding tests, which is a mistake imho since we never check for usage in typing.
I'd be interested in seeing adding `test_proxy_headers.py` makes a difference

Reply:
> I now realize we added the whole typing without the corresponding tests, which is a mistake imho since we never check for usage in typing.

I can see it as a plus, but not as a mistake.  

> I'd be interested in seeing adding test_proxy_headers.py makes a difference

I'm going to add it tonight then. Let's see. :+1:

Reply:
yep mistake is probably too strong, this said, I'm no mypy pro but this:
```
        if isinstance(trusted_hosts, str):
            self.trusted_hosts = {item.strip() for item in trusted_hosts.split(",")}
        else:
            self.trusted_hosts = set(trusted_hosts)
```
is pretty clear to me: `trusted_hosts` it's a tr or something else, so I dont know why `trusted_hosts: str = "127.0.0.1"` is valid here !
it's probably my bad understanding of mypy though but I'd love to undertstand :)

Reply:
> is pretty clear to me: trusted_hosts it's a tr or something else

Yeah, to me as well. I didn't pay enough attention. 

> it's probably my bad understanding of mypy though but I'd love to undertstand :)

`mypy` can't infer based on "future conditionals". I've declared the `trusted_hosts` to be `str`, so that's it. What we understand seeing this tho, is that `trusted_hosts` will always be `str`, then on the conditional, we'll always be on the first logic branch. This is not the case... 

Anyway, you're are right. Covering the test would have helped me to see this.

Reply:
I played a little bit with this, this is a hard one, I'll need more time